File: backend/src/hts_oracle/schemas/batch.py
"""
Pydantic schemas for batch classification.

Defines the SSE (Server-Sent Events) event types that the backend streams
to the frontend during batch PDF processing.

Each event type is a separate model so both backend and frontend have
a clear contract for what data each event carries.
"""

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# SSE event types — one model per event kind
# ---------------------------------------------------------------------------

class PhaseEvent(BaseModel):
    """Progress update: which phase we're in and overall progress."""
    event: str = "phase"
    phase: str              # "extracting_text", "extracting_commodities", "searching", "resolving"
    progress: int           # 0-100
    total: int = 0          # Total items (set after extraction)

class ItemProgressEvent(BaseModel):
    """One item has been processed (or is starting)."""
    event: str = "item_progress"
    index: int              # Which item (0-based)
    total: int              # Total items
    commodity: str          # Product description
    status: str             # "searching", "confident", "ambiguous"
    hts_code: str = ""      # Best match (empty if still searching)
    confidence: float = 0   # Confidence score 0-100

class CompleteEvent(BaseModel):
    """All items have been classified. Final results."""
    event: str = "complete"
    items: list[dict]       # Full classified items array
    summary: dict           # { total, classified, needs_review, avg_confidence }

class ErrorEvent(BaseModel):
    """Something went wrong during processing."""
    event: str = "error"
    message: str


# ---------------------------------------------------------------------------
# Batch job response (returned by upload endpoint)
# ---------------------------------------------------------------------------

class BatchUploadResponse(BaseModel):
    """Returned when a PDF is uploaded. Use job_id to connect to the SSE stream."""
    job_id: int
    filename: str
    message: str = "Upload successful. Connect to the stream endpoint for progress."

File: frontend/src/api/batch.ts
/**
 * Batch API client — handles PDF upload and SSE streaming.
 *
 * The batch flow is two steps:
 *   1. Upload PDF → get a job_id
 *   2. Connect to SSE stream → receive real-time progress events
 *
 * This module provides typed functions for both steps, plus a typed
 * event parser for the SSE stream.
 */

const API_BASE_URL = import.meta.env.VITE_API_URL ?? "";

// ---------------------------------------------------------------------------
// Types for SSE events (mirrors backend/schemas/batch.py)
// ---------------------------------------------------------------------------

export interface PhaseEvent {
  event: "phase";
  phase: string;
  progress: number;
  total: number;
}

export interface ItemProgressEvent {
  event: "item_progress";
  index: number;
  total: number;
  commodity: string;
  status: "searching" | "confident" | "ambiguous" | "needs_review";
  hts_code?: string;
  confidence?: number;
}

export interface CompleteEvent {
  event: "complete";
  items: BatchItem[];
  summary: BatchSummary;
}

export interface ErrorEvent {
  event: "error";
  message: string;
}

export type BatchSSEEvent = PhaseEvent | ItemProgressEvent | CompleteEvent | ErrorEvent;

export interface BatchItem {
  commodity: string;
  quantity?: string;
  value?: string;
  hts_code: string;
  description: string;
  confidence: number;
  general_rate: string;
  status: "confident" | "ambiguous" | "llm_assisted" | "needs_review";
}

export interface BatchSummary {
  total: number;
  classified: number;
  needs_review: number;
  avg_confidence: number;
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

/**
 * Upload a PDF for batch classification. Returns a job_id.
 */
export async function uploadPdf(file: File): Promise<{ job_id: number; filename: string }> {
  const formData = new FormData();
  formData.append("file", file);

  const response = await fetch(`${API_BASE_URL}/api/v1/batch/upload`, {
    method: "POST",
    body: formData,
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error((error as Record<string, string>).detail ?? `Upload failed (${response.status})`);
  }

  return response.json();
}

/**
 * Connect to the SSE stream for a batch job.
 *
 * This is the SSE client. It reads the stream and calls the onEvent
 * callback for each parsed event. The callback receives typed events
 * so the component doesn't need to do any parsing.
 *
 * Returns an AbortController — call .abort() to disconnect.
 */
export function streamBatchProgress(
  jobId: number,
  onEvent: (event: BatchSSEEvent) => void,
  onError: (error: Error) => void,
): AbortController {
  const controller = new AbortController();

  // Use fetch + ReadableStream instead of EventSource.
  // EventSource doesn't support custom headers or POST requests,
  // and it auto-reconnects in ways we don't want to handle here.
  (async () => {
    try {
      const response = await fetch(`${API_BASE_URL}/api/v1/batch/${jobId}/stream`, {
        signal: controller.signal,
      });

      if (!response.ok) {
        throw new Error(`Stream connection failed (${response.status})`);
      }

      const reader = response.body!.getReader();
      const decoder = new TextDecoder();
      let buffer = "";

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });

        // SSE events are separated by double newlines
        const parts = buffer.split("\n\n");
        buffer = parts.pop()!; // Keep incomplete chunk

        for (const part of parts) {
          const line = part.trim();
          if (!line.startsWith("data: ")) continue;

          try {
            const event = JSON.parse(line.slice(6)) as BatchSSEEvent;
            onEvent(event);
          } catch {
            // Skip malformed events
          }
        }
      }
    } catch (err) {
      if ((err as Error).name !== "AbortError") {
        onError(err as Error);
      }
    }
  })();

  return controller;
}